src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs

```diff
@@ -0,0 +1,193 @@
+﻿namespace ServiceControl.Transports.SQLServer
+{
+	using System;
+	using System.Collections.Concurrent;
+	using System.Collections.Generic;
+	using System.Data.SqlClient;
+	using System.Linq;
+	using System.Threading;
+	using System.Threading.Tasks;
+	using Monitoring;
+	using Monitoring.Infrastructure;
+	using Monitoring.Messaging;
+	using Monitoring.QueueLength;
+	using NServiceBus.Logging;
+	using NServiceBus.Metrics;
+
+	public class QueueLengthProvider : IProvideQueueLength
+	{
+		ConcurrentDictionary<EndpointInputQueue, SqlTable> endpointInputQueueToTableName = new ConcurrentDictionary<EndpointInputQueue, SqlTable>();
+		ConcurrentDictionary<SqlTable, int> tableSizes = new ConcurrentDictionary<SqlTable, int>();
+
+		string connectionString;
+		QueueLengthStore store;
+
+		CancellationTokenSource stop = new CancellationTokenSource();
+		Task pooler;
+
+		public void Initialize(string connectionString, QueueLengthStore store)
+		{
+			this.connectionString = connectionString;
+			this.store = store;
+		}
+
+		public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+		{
+			var endpointInputQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+			var localAddress = metadataReport.LocalAddress;
+			var pluginVersion = metadataReport.PluginVersion;
+
+			if (SqlTable.TryParse(localAddress, pluginVersion, out var sqlTable))
+			{
+			    SqlTable previousSqlTable = null;
+
+				endpointInputQueueToTableName.AddOrUpdate(endpointInputQueue, _ => sqlTable, (_, cv) =>
+				{
+					previousSqlTable = cv;
```

Comment:
I think that now I understand how it works. I'd propose anonther approach. For the `(_, cv)` function, we could immiediately remove the tableSize `tableSize.TryRemove (previousSqlTable)` and just add the new value below without any condition. This, IMO, would make it easier to understand